recipes: Add company-coq

cpitclaudel opened the pull request:
company-coq is a company back-end that offers documentation and keywords
autocompletion when editing Coq proof scripts.

I've tested this as directed in the manual, and everything seems to work
great on Emacs 24.4. I had a bit more trouble when testing the package
on 24.3 in an Ubuntu VM; the dependencies were compiled in incorrect
order (for some reason company-math was compiled first, so the compilation of
company-math failed when it tried to require company; similarly,
company-coq failed when it tried to require company-math, as
company-math had already failed). Did I do anything wrong there?

purcell commented:
Thanks for this. Some feedback:

The `:branch` setting in the recipe is inappropriate -- MELPA deliberately builds packages from the latest development source code, usually `master`. For stable releases, place a version tag wherever you like and MELPA Stable will build a package from it. You can amend the recipe commit and force-push it here to update the pull request cleanly.

Also, rather than include the LICENSE file, just place the standard GNU preamble directly in the .el file, as `auto-insert-mode` does when you create a fresh .el file with it enabled.

Additionally, `:defaults` is preferable to `*.el`, since it is a little more future-proof.

The `-pkg.el` file should list the *stable* versions of the dependency packages: you can find these on stable.melpa.org.

You should usually not make a `company` plug-in depend on `yasnippet`: not all `company` users want to have `yasnippet` installed. Better, if possible, to make this an optional dependency by placing `yasnippet`-related code in an `(eval-after-load ...)` block or a separate `company-coq-yasnippet.el` file.

Please place a short description in the `;;; Commentary:` header of the main `.el` file: this text will appear on the MELPA package package, and in `M-x list-packages` when the user presses `?`.

You're using `setq-local`, which implies you require Emacs 24.3 or greater. You probably just want to use `(set (make-local-variable 'name) val)` instead.

Hope that helps!

-Steve

purcell commented:
P.S. I'm not sure why you saw problems with the compilation order in Emacs 24.3 -- don't worry about that too much. I don't think you're doing anything wrong.

cpitclaudel commented:
Thanks for the feedback! That's pretty cool. I've updated the PR to implement your suggestions:

* Added a desc and the license in the file's header
* Removed LICENSE and replaced "*.el" with :defaults in the recipe.
* Replaced uses of setq-local
* Removed the :branch spec
* Changed the version numbers to the relevant stable ones

A few questions:

* Once (if!) the package gets in MELPA, is it still ok to rewrite git history? (i.e. force push things), or will that break MELPA?

* `company-math` doesn't seem to have a stable version. What should I do?

Regarding Yasnippet: that's a tricky one. The package really doesn't do much at all without Yasnippet: A large majority of the completions are actually Yasnippet patterns; the idea is that commands ("tactics" in Coq parlance) are lengthy (for example `dependent inversion_clear @{ident} as @{intropattern} with @{term}`) and most of them have "holes" that the user should fill. Moth of the completion that company-coq does is auto-inserting these snippets after the user types a few characters. Without Yasnippet this would be totally useless. Proof-general even has basic support for that via the included 'Holes' feature, but its author agrees that it would be nice to move to a more actively developed alternative.

If there were more things to complete that were not patterns with holes I'd agree with removing the Yasnippet dependency, but as the package stands I think it wouldn't be too useful without Yasnippet. Plus, given the particular dependencies of this package (it requires Proof-General, which itself is over a megabyte), I don't think Yasnippet represents a huge additional burden.

What do you think?

purcell commented:
> Thanks for the feedback! That's pretty cool. I've updated the PR to implement your suggestions:
> 
>     Added a desc and the license in the file's header
>     Removed LICENSE and replaced "*.el" with :defaults in the recipe.
>     Replaced uses of setq-local
>     Removed the :branch spec
>     Changed the version numbers to the relevant stable ones
> 

Great, thanks.


> A few questions:
> 
>     Once (if!) the package gets in MELPA, is it still ok to rewrite git history? (i.e. force push things), or will that break MELPA?

It won't break MELPA, but if the git repo changes while the timestamp of the last commit remains the same, then the corresponding MELPA package will get silently updated with new contents: the version of the package won't show that the contents changed. In general, avoid force-pushing if you can.

>     company-math doesn't seem to have a stable version. What should I do?

Use `"0"` in this case please.

> 
> Regarding Yasnippet: that's a tricky one. The package really doesn't do much at all without Yasnippet:
> ...
> What do you think?

That's fine, then.

cpitclaudel commented:
Awesome, thanks a lot! And thanks for the great feedback (and of course for your work on MELPA!)

cpitclaudel commented:
I've updated the version number for company-math. Thanks again!

purcell commented:
Happy to help. Thanks for the submission and for fixing everything up so quickly.
